File: 2021/day1.erl
#!/usr/bin/env escript
%% -*- erlang -*-

main(_) ->
    {ok, File} = file:open("day1.txt", [read]),
    Lines = read_all_ints(File),
    Part1 = count_depth_increases(Lines, 1),
    Part2 = count_depth_increases(Lines, 3),
    io:format("~B ~B~n", [Part1, Part2]).

count_depth_increases(Depths, WindowSize) ->
    count_depth_increases(Depths, lists:nthtail(WindowSize, Depths), 0).

count_depth_increases(_, [], N) ->
    N;
count_depth_increases([Expire | Old], [Next | New], N) ->
    Deeper = if Next > Expire -> 1; true -> 0 end,
    count_depth_increases(Old, New, N + Deeper).

read_all_ints(File) ->
    lists:reverse(read_all_ints(File, [])).

read_all_ints(File, Acc) ->
    case file:read_line(File) of
        {error, _} -> Acc;
        eof -> Acc;
        {ok, Line} -> read_all_ints(File, [list_to_integer(string:trim(Line)) | Acc])
    end.

File: 2024/day1/day1.pl
#!/usr/bin/perl

use strict;

my (@a, @b, %b);

while (<>) {
    my ($a, $b) = split;
    push @a, $a;
    push @b, $b;
    $b{$b}++;
}

@a = sort { $a <=> $b } @a;
@b = sort { $a <=> $b } @b;

my ($part1, $part2);
for my $i (0..$#a) {
    $part1 += abs($a[$i] - $b[$i]);
    $part2 += $a[$i] * $b{$a[$i]}
}

print "$part1 $part2\n";
